xarray/tests/test_dataset.py

```diff
@@ -3431,6 +3431,36 @@ def test_expand_dims_kwargs_python36plus(self) -> None:
         )
         assert_identical(other_way_expected, other_way)
 
+    @pytest.mark.parametrize("create_1d_index_flag", [True, False])
+    def test_expand_dims_create_index_data_variable(self, create_1d_index_flag):
+        # data variables should not gain an index ever
+        ds = Dataset({"x": 0})
+        expanded = ds.expand_dims("x", create_1d_index=create_1d_index_flag)
+
+        # TODO I can't just create the expected dataset directly using constructor because of GH issue 8959
+        # expected = Dataset(data_vars={"x": ("x", [0])})
+        expected = Dataset({"x": ("x", [0])}).drop_indexes("x").reset_coords("x")
```

Comment:
See #8959

Reply:
I didn't think we should merge without either updating the invariants or fixing the behavior or test

Reply:
Yeah I agree, I wouldn't really have felt comfortable merging this at this point either.

Reply:
So the action here is to
1) Use the [private version of the asserts](https://github.com/pydata/xarray/blob/b0036749542145794244dee4c4869f3750ff2dee/xarray/tests/__init__.py#L286), which have a `check_default_indexes` kwarg
2) We should probably publicly expose that kwarg in the public assert functions API
3) We should write a little documentation explaining what it does, to go in the [docs page on testing](https://docs.xarray.dev/en/stable/user-guide/testing.html#testing-your-code)

Reply:
The private version *should* be getting used in all the tests anyway.

I don't know about exposing it publicly. It's our job to make sure they're always satisfied when using the API.

Reply:
> The private version should be getting used in all the tests anyway.

+1

> I don't know about exposing it publicly. It's our job to make sure they're always satisfied when using the API.

Isn't that only true if indexes are a hidden private part of the data model? If they are public, users should have control over them (including whether or not they exist for any given 1D dimension coordinate).

Perhaps we should not be exposing the kwarg, but instead setting `check_default_indexes=False` on the public asserts in favour of changing them to actually compare all the indexes and report if the indexes are unequal. i.e. what #5812 is suggesting.

Reply:
> hanging them to actually compare all the indexes and report if the indexes are unequal.

This makes sense to me.

Reply:
Fixed in https://github.com/pydata/xarray/pull/8960/commits/62e750f46c55209b904cc14ece174fa58cce546a by using private versions of assertions. The rest of this discussion is already covered by #5812.